feat: update the tailwind plugin to suuport tailwindv4 #2819

SisyphusZheng opened the pull request:
Inspired by #2819, #2829, and https://github.com/pakornv/fresh-plugin-tailwindcss, I made minimal modifications to the plugin-tailwind, I originally intended to update based on #2829, but I was informed by the system that I couldn't push.

 Due to the breaking changes in Tailwind v4, some PostCSS plugins are no longer required. Additionally, I updated the syntax in the /www directory. The /www directory now uses the updated plugin supported by Tailwind v4.I think this has made initial progress.

# What This PR Solves
The Tailwind plugin has been updated to support Tailwind v4, and it is currently functioning well both in the repository's www directory and in the initialized project which for test.

## --patch-1

### 1. Dependency Optimization
- Removed `cssnano` dependency, using Tailwind v4's built-in optimization(Based on @csvn  's suggestion)
- Removed `autoprefixer` dependency
- Updated `tailwind-plugin` version number
### 2. Build Configuration Adjustments
- Added `--node-modules-dir` parameter to `/www` task for Tailwind v4 support (Based on @marvinhagemeister 's suggestion)
- Added `--allow-scripts` parameter to enable `@tailwindcss/oxide` execution
### 3. Code Adaptation
- Modified Tailwind-related code in `/www` to support Tailwind v4
- Updated dependency configuration in `deno.json`
### 4. Feature Restoration
- Added `imagescript` and `astral` dependencies to restore screenshot functionality lost in previous PR updates
- Fixed the issue where the `tailwind-plugin` dependency was incorrectly pointing to the JSR package instead of the local repository source code in previous PR.

## --patch-2
- Upgraded Tailwind CSS to the latest version.
- Removed unnecessary dependencies.
- Enabled CSS minification by default when Fresh is in production mode.
- Updated docs/canary/examples/migration-guide.md to support the new Tailwind plugin format.
- Updated init/src/init.ts to support the new Tailwind version and plugin version, use new `@import "tailwindcss"`, and generate the new plugin format.
- Updated init/src/init_test.ts: tests now expect `css` to match `/tailwindcss/` instead of `/@tailwind/`.
- Removed plugin-tailwindcss/src/types.ts as it is no longer needed, and updated plugin-tailwindcss/src/mod.ts and www/dev.ts to support the new format.

Co-authored-by: Asher Gomez <ashersaupingomez@gmail.com>

## --patch-3
Addressing @marvinhagemeister  feedback, this patch reverts the unintended loss of PR #2960's changes caused by merge conflicts.

## --patch-4

I initialized a project according to the migration documentation and used @fresh/update to test whether the plugin works properly with v2-alpha-34. The results show that it works fine. 

- The plugin is published at jsr:@freshpress/fresh-plugin-tailwind@0.0.1-alpha.8. 

- The test project is published at https://github.com/SisyphusZheng/fresh-project-test/tree/main/fresh-project. 

## --patch-5

As plugin, work fine with fresh v2 proj, can check: [fresh-tailwindv4-template](https://github.com/SisyphusZheng/fresh-project-test/tree/main/fresh-tailwindv4-template)

# How Testing Was Conducted

## Local Environment Testing
1. Environment Cleanup
```bash
rm -rf ./.cache/deno deno.lock node_modules
```

2. Test Execution
```bash
deno task www
```
3. Result Verification
- Checked functionality in local `/www` environment
- Confirmed all features working properly

## Codespace Environment Testing
- This forked /www website is hosted and running on GitHub Codespaces, with support for the latest code changes. You can check it out at:
https://silver-trout-pjr4jw7xx5x9299x6-8000.app.github.dev/

## As plugin for New Fresh v2 proj Environment Testing
- As plugin, work fine with fresh v2 proj, can check: [fresh-tailwindv4-template](https://github.com/SisyphusZheng/fresh-project-test/tree/main/fresh-tailwindv4-template)

SisyphusZheng committed:
feat: plugin support tailwindv4

deno-deploy[bot] committed:
[Deno Deploy] Update .github/workflows/deploy.yml

SisyphusZheng committed:
Update deploy.yml

SisyphusZheng committed:
Update deploy.yml

deno-deploy[bot] committed:
[Deno Deploy] Update .github/workflows/deploy.yml

SisyphusZheng committed:
Merge branch 'main' of https://github.com/SisyphusZheng/support-tailwindv4

SisyphusZheng committed:
Merge branch 'main' of https://github.com/SisyphusZheng/support-tailwindv4

SisyphusZheng committed:
Update deploy.yml

SisyphusZheng committed:
fix format for pass deno ok

SisyphusZheng committed:
Update tailwind.config.ts

deno-deploy[bot] committed:
[Deno Deploy] Update .github/workflows/deploy.yml

SisyphusZheng committed:
Update deploy.yml

SisyphusZheng committed:
Update deploy.yml

SisyphusZheng committed:
Update deploy.yml

csvn commented on plugin-tailwindcss/src/compiler.ts:
```diff
@@ -1,80 +1,23 @@
-    plugins.push(cssnano());
+    // deno-lint-ignore no-explicit-any
+    plugins.push(cssnano(options.cssnanoOptions ?? {}) as any);
   }
```
I'm pretty sure `cssnano` can be removed in favor of passing an [option to `@tailwind/postcss`](https://github.com/tailwindlabs/tailwindcss/blob/main/packages/%40tailwindcss-postcss/src/index.ts#L55):

```ts
const minify = config.mode === "production" && options.minify !== false;
twPostcss({
  optimize: minify ? { minify: true } : true // or `false` for no optimizations?
});
```

I think it's better to pull in as few dependencies as possible (only `tailwindcss`, `@tailwindcss/postcss` & `postcss`), and leave it to PostCSS + built-in LightningCSS.

Looking forward to having v4 version though!

SisyphusZheng committed:
Merge branch 'main' of https://github.com/SisyphusZheng/fresh

SisyphusZheng committed:
feat(tailwind): initial support for Tailwind v4 plugin with working local implementation

SisyphusZheng committed:
fix: pass the deno ok

iuioiua commented on deno.json:
```diff
@@ -65,27 +65,25 @@
     "fresh/runtime": "./src/runtime/shared.ts",
     "github-slugger": "npm:github-slugger@^2.0.0",
+    "imagescript": "https://deno.land/x/imagescript@1.3.0/mod.ts",
+    "astral": "jsr:@astral/astral",
```
The `@astral/astral` import already exists above.
```suggestion
```

iuioiua commented on deno.json:
```diff
@@ -65,27 +65,25 @@
-    "postcss": "npm:postcss@8.4.35",
-    "tailwindcss": "npm:tailwindcss@^3.4.1",
-
+    "postcss": "npm:postcss@8.5.3",
```
So Tailwind-related imports are grouped together.
```suggestion

    "@tailwindcss/postcss": "npm:@tailwindcss/postcss@^4.1.7",
    "@tailwindcss/oxide": "npm:@tailwindcss/oxide@^4.1.7",
    "postcss": "npm:postcss@8.5.3",
```

iuioiua commented on deno.json:
```diff
@@ -65,27 +65,25 @@
     "@std/streams": "jsr:@std/streams@1",
-
+    "@tailwindcss/postcss": "npm:@tailwindcss/postcss@^4.1.7",
+    "@tailwindcss/oxide": "npm:@tailwindcss/oxide@^4.1.7",
```
See below
```suggestion
```

iuioiua commented on deno.json:
```diff
@@ -65,27 +65,25 @@
-    "tailwindcss": "npm:tailwindcss@^3.4.1",
-
+    "postcss": "npm:postcss@8.5.3",
+    "tailwindcss": "npm:tailwindcss@^4.0.0",
```
Latest version. Are the other imports the latest versions?
```suggestion
    "tailwindcss": "npm:tailwindcss@^4.1.7",
```

iuioiua commented on plugin-tailwindcss/src/types.ts:
```diff
@@ -1,48 +1,19 @@
+   * Tailwind CSS configuration
    */
-  overrideBrowserslist?: string | string[];
+  config?: Config;
```
Is there an API reference we can `{@link}` to?

iuioiua commented on plugin-tailwindcss/src/types.ts:
```diff
@@ -1,48 +1,19 @@
+  /**
+   * PostCSS options
+   */
+  postcssOptions?: Record<string, unknown>;
```
Ditto. And is there an interface we can use here?

SisyphusZheng replied:
> Ditto. And is there an interface we can use here?

Based on your suggestion, I have re-reviewed the project and think that there is no reason to use type.ts. The compression functionality can be directly integrated into the compiler for implementation.types

iuioiua commented on plugin-tailwindcss/src/types.ts:
```diff
@@ -1,48 +1,19 @@
-  exclude?: OnTransformOptions["exclude"];
+  /**
+   * Whether to minify the CSS in production mode
+   * @default true
```
```suggestion
   * Whether to minify the CSS. If Fresh is in production mode (`config.mode === "production"`) and this undefined, this is automatically set to `true`.
   *
   * @default {true}
```

iuioiua commented on www/deno.json:
```diff
@@ -1,6 +1,6 @@
 {
   "tasks": {
-    "start": "deno run -A --watch=static/,routes/,../src,../docs dev.ts",
+    "start": "deno run -A --allow-scripts --node-modules-dir --watch=static/,routes/,../src,../docs dev.ts",
```
Might it be better to add `"nodeModulesDir": "auto"` to this file?

SisyphusZheng replied:
> Might it be better to add `"nodeModulesDir": "auto"` to this file?

The awkward part is that nodeModulesDir can only be added in the root directory. If I do this, I'll encounter a Preact error message, and I'm not sure if it will cause more issues. For now, adding --allow-scripts --node-modules-dir only in /www allows it to run relatively smoothly.

SisyphusZheng committed:
Apply suggestions from @iuioiua code review

## Here is a simple description of what this update did:
1.Upgraded Tailwind CSS to the latest version.
2.Removed unnecessary dependencies.
3.Enabled CSS minification by default when Fresh is in production mode.

Co-authored-by: Asher Gomez <ashersaupingomez@gmail.com>

SisyphusZheng committed:
refactor: del type.ts and update the init.ts

iuioiua reviewed (approved):
Great work!

marvinhagemeister commented on plugin-tailwindcss/src/mod.ts:
```diff
@@ -1,19 +1,17 @@
 
   builder.onTransformStaticFile(
-    { pluginName: "tailwind", filter: /\.css$/, exclude: options.exclude },
+    { pluginName: "tailwind", filter: /\.css$/ },
```
Maybe not intentional: This change here removes the ability to exclude files from being processed by tailwind.

SisyphusZheng replied:
Yes, because the Exclude feature hadn't been implemented in the previous Fork.

marvinhagemeister commented on plugin-tailwindcss/README.md:
```diff
@@ -1,6 +1,6 @@
 # Tailwind CSS plugin for Fresh
 
-A Tailwind CSS plugin to use in Fresh.
+A Tailwind CSS v4 plugin to use in Fresh.
```
Nit: imo no need to say that it's v4. People always assume that it's latest.

SisyphusZheng committed:
Update README.md

SisyphusZheng committed:
Update mod.ts

SisyphusZheng committed:
fix: Restore the functionality of the plugin as it was in update #2960.

marvinhagemeister committed:
m

marvinhagemeister committed:
Merge branch 'main' into pr-tailwind-v4

marvinhagemeister committed:
fix: nodeModulesDir handling with tailwind

marvinhagemeister committed:
update tw plugin docs

marvinhagemeister committed:
update lock

marvinhagemeister reviewed (approved):
LGTM, thanks for doing this, this is great!

marvinhagemeister committed:
don't run tests in parallel on CI

marvinhagemeister committed:
try to fix ci

marvinhagemeister committed:
fix fmt

marvinhagemeister committed:
Revert "try to fix ci"

This reverts commit 61c02cac2332232625c545e2a3546584a398c2ca.

marvinhagemeister committed:
Revert "fix: nodeModulesDir handling with tailwind"

This reverts commit 6d72efea3a4dd36ff2bc3efe87b6f529ce6bce3d.

marvinhagemeister committed:
Revert "don't run tests in parallel on CI"

This reverts commit 20302213b731eae3c099adcf626816a914ba9530.
